applications/FluidDynamicsApplication/custom_utilities/mass_conservation_utility.cpp

```diff
@@ -246,75 +317,97 @@ void MassConservationCheckProcess::ComputeVolumesAndInterface( double& positiveV
-            // all nodes are on one side and the element is not split
-            continue;
-        }
-        else {
+        const bool geom_is_cut = IsGeometryCut(r_geom, pt_count_neg, pt_count_pos);
+        if (geom_is_cut){
             // element is cut by the surface (splitting)
-            Kratos::unique_ptr<ModifiedShapeFunctions> p_modified_sh_func = nullptr;
             Vector w_gauss_interface(3, 0.0);
 
-            Vector Distance( rGeom.PointsNumber(), 0.0 );
-            for (unsigned int i = 0; i < rGeom.PointsNumber(); i++){
-                Distance[i] = rGeom[i].FastGetSolutionStepValue(DISTANCE);
+            Vector distance( r_geom.PointsNumber(), 0.0 );
+            for (unsigned int i = 0; i < r_geom.PointsNumber(); i++){
+                // Control mechanism to avoid 0.0 ( is necessary because "distance_modification" possibly not yet executed )
+                double &r_dist = r_geom[i].FastGetSolutionStepValue(DISTANCE);
+                const double min_distance = r_geom[i].GetValue(NODAL_H)*1e-6;
+                if ( std::abs(r_dist) < min_distance ){
+                    r_geom[i].SetLock();
+                    r_dist = r_dist > 0 ? min_distance : -min_distance;
+                    r_geom[i].UnSetLock();
+                }
+                distance[i] = r_dist;
             }
 
-            if ( rGeom.PointsNumber() == 3 ){ p_modified_sh_func = Kratos::make_unique<Triangle2D3ModifiedShapeFunctions>(it_elem->pGetGeometry(), Distance); }
-            else if ( rGeom.PointsNumber() == 4 ){ p_modified_sh_func = Kratos::make_unique<Tetrahedra3D4ModifiedShapeFunctions>(it_elem->pGetGeometry(), Distance); }
-            else { KRATOS_ERROR << "The process can not be applied on this kind of element" << std::endl; }
+            const auto p_modified_sh_func = GetModifiedShapeFunctions(it_elem->pGetGeometry(), distance);
 
             // Concerning their area, the positive and negative side of the interface are equal
-            p_modified_sh_func->ComputeInterfacePositiveSideShapeFunctionsAndGradientsValues(
+            p_modified_sh_func->ComputeInterfaceNegativeSideShapeFunctionsAndGradientsValues(
                     shape_functions,                    // N
                     shape_derivatives,                  // DN
                     w_gauss_interface,                  // includes the weights of the GAUSS points (!!!)
-                    GeometryData::GI_GAUSS_1);          // first order Gauss integration
+                    GeometryData::GI_GAUSS_2);          // second order Gauss integration
+
+            // negative side outwards area normal vector values for the Gauss pts. of given quadrature
+            std::vector<Vector> normal_vectors;
+            p_modified_sh_func->ComputeNegativeSideInterfaceAreaNormals(
+                normal_vectors,
+                GeometryData::GI_GAUSS_2
+            );
+
+            // iteration over all 3 or 6 integration points
```

Comment:
```suggestion
            // iteration over interface integration points
```
This is not correct.